EIPS/eip-7747.md

```diff
@@ -0,0 +1,242 @@
+
+Opcodes `ADDMODX(0xc3)`, `SUBMODX(0xc4)`, `MULMODX(0xc5)` take a 3-byte immediate interpreted as 3 values `out`, `x`, `y`.  These are indices which correspond to individual values from the space of the active field context (0 corresponds to the first value in the space, 1 to the second, etc.).
+
+Execution asserts that an active field context is set in the current call frame, asserts that each input index is less than the size of the active field context's value space.
+
+Then, compute the given operation with inputs from slot `x`/`y`:
+
+* `ADDMODX`: `value_space[x] + value_space[y] % modulus`
+* `SUBMODX`: `value_space[x] - value_space[y] % modulus`
+* `MULMODX`: `value_space[x] * value_space[y] % modulus` 
+    
+Store the result in the value at index `out`.  Note: `x`, `y` and `out` are allowed to overlap.
+
+Charge a gas cost based on `active_context.element_size/8` (the number of 64bit words needed to represent values/modulus):
+
+```
+# element_size64bit \in [1, 12]
+
+def gas_mulmodx(element_size64bit: int) -> int:
+    coef_a = 2.3
+    coef_b = -2
+    coef_c = 32
+    return math.ceil(coef_a * element_size64bit ** 2 + \
+        coef_b * element_size64bit + coef_c)  
+    
+def gas_addmodx(element_size64bit: int) -> int:
+    coef_a = 0.1111
+    coef_b = 1.111
+    return math.ceil(coef_a * element_size64bit + coef_b)
+    
+def gas_setmod(element_size64bit: int) -> int:
+    coef_a = 110
+    coef_b = 730
+    return math.ceil(coef_a * element_size64bit + coef_b)
+```
+
+`SUBMODX` uses the same cost model as `ADDMODX`.
+
+#### Cost table:
+
+| Modulus Size | MULMODX cost | ADDMODX and SUBMODX cost | SETMODX cost |
+| ---- | ---- |  ---- | ---- |
+|1 - 64 bits|2|2|32|
+|65 - 128 bits|2|2|36|
+|129 - 192 bits|2|2|40|
+|193 - 256 bits|3|2|44|
+|257 - 320 bits|3|2|48|
+|321 - 384 bits|4|2|52|
+|385 - 448 bits|5|2|56|
+|449 - 512 bits|7|2|60|
+|513 - 576 bits|8|3|64|
+|577 - 640 bits|9|3|68|
+|641 - 704 bits|11|3|72|
+|705 - 768 bits|13|3|76|
+
+--- 
+
+### Data Transfer
+
+#### `LOADX(0xc1)`
+
+**Stack in**: `(top of stack) dest source count`
+
+**Stack out**: none
+
+**Description**: copies a number of contiguous values in the currently-active field context value space into EVM memory.
+
+##### Execution
+
+* Assert that a field context is set as active in the current call frame.
+* Assert that the range `[source, source + count]` falls within the active field context value space.
+* Assert that the destination of the copy falls entirely within EVM memory.
+* Charge `gas_mulmodx(active_context.element_size / 8) * count` and copy the values into EVM memory starting at offset `dest`, laying them out in big-endian ordering, padded such that size of each is a multiple of 64 bits.
+
+#### `STOREX(0xc2)`
+
+**Stack in**: `dest source count`
+
+**Stack out**: none
+
+**Description**: copies values layed out in EVM memory as 64bit-aligned big-endian numbers into a contiguous section of the currently-active field context's value space.
```

Comment:
It is unclear if the layout is compatible with BLS12-381 (i.e. EIP-2537) layout, I think it is but padding should be mentioned